dockstore-webservice/src/main/java/io/dockstore/webservice/languages/CWLHandler.java

```diff
@@ -896,4 +899,271 @@ public VersionTypeValidation validateToolSet(Set<SourceFile> sourcefiles, String
+     * will throw an exception.
+     */
+    public static class Preprocessor {
+        private static final List<String> IMPORT_KEYS = Arrays.asList("$import", "import");
+        private static final List<String> INCLUDE_KEYS = Arrays.asList("$include", "include");
+        private static final List<String> MIXIN_KEYS = Arrays.asList("$mixin", "mixin");
+        private static final String EMPTY_STRING = "";
+        private static final Map<String, Object> EMPTY_MAP = Collections.emptyMap();
+        private static final int DEFAULT_MAX_DEPTH = 10;
+        private static final long DEFAULT_MAX_CHAR_COUNT = 4L * 1024L * 1024L;
+        private static final long DEFAULT_MAX_FILE_COUNT = 2000L;
+
+        private final Set<SourceFile> sourceFiles;
+        private final Map<String, String> idToPath;
+        private long charCount;
+        private long fileCount;
+        private final int maxDepth;
+        private final long maxCharCount;
+        private final long maxFileCount;
+
+        /**
+         * Create a CWL Preprocessor with specified "max" values.  See the class javadoc regarding the meaning of the "max" arguments.
+         * @param sourceFiles files to search when expanding $import, $include, etc
+         * @param maxDepth the maximum file depth
+         * @param maxCharCount the maximum number of expanded characters (approximate)
+         * @param maxFileCount the maximum number of files expanded
+         */
+        public Preprocessor(Set<SourceFile> sourceFiles, int maxDepth, long maxCharCount, long maxFileCount) {
+            this.sourceFiles = sourceFiles;
+            this.idToPath = new HashMap<>();
+            this.charCount = 0;
+            this.fileCount = 0;
+            this.maxDepth = maxDepth;
+            this.maxCharCount = maxCharCount;
+            this.maxFileCount = maxFileCount;
+        }
+
+        /**
+         * Create a CWL Preprocessor with default "max" values.
+         */
+        public Preprocessor(Set<SourceFile> sourceFiles) {
+            this(sourceFiles, DEFAULT_MAX_DEPTH, DEFAULT_MAX_CHAR_COUNT, DEFAULT_MAX_FILE_COUNT);
+        }
+
+        /**
+         * Preprocess the specified CWL, recursively expanding various directives as noted in the class javadoc.
+         * This method may, but does not necessarily, process the specified CWL in place.
+         * @param cwl a representation of a CWL file or portion thereof, typically a Map, List, or String and the result of new Yaml().load(content)
+         * @param currentPath the path of the file containing the CWL represented by obj
+         * @param depth the current file depth, where the root file is at depth 0, and the depth is incremented on each file expansion
+         * @return the preprocessed CWL
+         */
+        public Object preprocess(Object cwl, String currentPath, int depth) {
+
+            if (depth > maxDepth) {
+                handleMax("maximum file depth exceeded");
```

Comment:
There are many Code Coverage warnings about not being covered by tests, including this one.

If the warnings are correct (I think sometimes it's flaky), it would be good to get better coverage for this file.

Reply:
There is a new unit test file that should cover most of `Preprocessor`.  Maybe the new file wasn't included in the analysis?  I believe the rest is fairly well-covered by the existing integration tests.

Reply:
I thought of some additional tests I can add to increase coverage.  Will add them a try to determine coverage on `Preprocessor` locally.